PyQt5, PyQt6, PySide6 Shim (Switch between different Python bindings for Qt versions)

jonoomph opened the pull request:
Related to https://github.com/OpenShot/libopenshot/pull/1027

  ### openshot-qt runtime

  - openshot-qt uses the env var `OPENSHOT_QT_API`.
  - Values:
      - `auto`: try bindings in order and use the first one that imports
      - `pyqt6`: *requires libopenshot to use Qt6*
      - `pyside6`:  *requires libopenshot to use Qt6*
      - `pyqt5`
  - In auto mode, it tries: `pyqt6`, then `pyside6`, then `pyqt5`.

  ### libopenshot build-time

  - libopenshot uses the CMake cache variable `-DUSE_QT6`.
  - Values:
      - `AUTO`: prefer Qt6 if found, otherwise fall back to Qt5
      - `ON`: force Qt6
      - `OFF`: force Qt5
  - After that, CMake sets QT_VERSION_MAJOR to 6 or 5 and links Qt${QT_VERSION_MAJOR}.

jonoomph committed:
Replacing QRegExp with QRegularExpression for Qt6 compatibility in mind. Also, adding a new QT version shim, although we aren't using it quite yet.

jonoomph committed:
Initial integration of qt_api shim, to switch between different versions of PyQt5 and PyQt6. VERY Experimental.

jonoomph commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Support PyQt5, PyQt6, PySide6 Shim (Switch between different Python bindings for Qt versions)](https://trello.com/c/XEGaTv2E/204-support-pyside-for-qt6-shim-to-easily-toggle-between-pyqt6-and-pyside-without-breaking-pyqt5)

jonoomph committed:
Initial PyQt6 support in our qt_api shim - tested in Ubuntu 25.04 VM. Both PyQt5 and PyQt6 are now working, except some cursor issues, which seems to get stuck after a drag/drop item operation.

jonoomph committed:
Initial support for PySide6, tested on Ubuntu 25.04 VM.

jonoomph committed:
Removing mentions of pyside2 (never supported)

jonoomph committed:
Updating develoers docs with PyQt5, PyQt6, and PySide6 packages and info.

jonoomph committed:
Merge branch 'develop' into qt6-support

# Conflicts:
#	src/windows/add_to_timeline.py
#	src/windows/animated_title.py
#	src/windows/export.py
#	src/windows/main_window.py
#	src/windows/models/emoji_model.py
#	src/windows/models/files_model.py
#	src/windows/models/titles_model.py
#	src/windows/models/transition_model.py
#	src/windows/profile_edit.py
#	src/windows/views/effects_listview.py
#	src/windows/views/emojis_listview.py
#	src/windows/views/files_listview.py
#	src/windows/views/properties_tableview.py
#	src/windows/views/timeline_backend/qwidget/track.py
#	src/windows/views/titles_listview.py
#	src/windows/views/transitions_listview.py

jonoomph committed:
Fixing merge issue, after merging develop back into this branch

jonoomph committed:
Fixing many new PyQt5/PyQt6 issues after merging in develop branch (this might take a few more tries)

jonoomph committed:
Fixing a few missing PyQt6 imports in qt_api

jonoomph committed:
Fixing some regex issues with PyQt6, preventing filtering from working on a few models (transitions, effects, files)

jonoomph committed:
Fixing a few minor PyQt6 errors and attempting to fix a minor focus/tab order issue.

jonoomph committed:
Fixing PyQt5 hard-coded import in focus tabstops code, which breaks PyQt6.

jonoomph committed:
Replacing more hard-coded PyQt5 imports with qt_api shims.

jonoomph committed:
Replaced all explicit Qt base-class __init__ calls with super().__init__() across the codebase.

jonoomph committed:
Fixing a Preferences error for PySide6 - causing the window to fail to appear.

jonoomph committed:
Fixing title bar height and margins for PySide6 compatibility (and for a consistent look on all widget libraries).

jonoomph committed:
Fixing tutorial dialog for better support for PyQt6 & PySide6 (importing QPointF correctly, ending QPainter correctly)

jonoomph committed:
Another fix to Tutorial dialog to use associatedWidgets for PyQt6 & PySide6 compatibility

jonoomph committed:
Improving Tutorial dialog to be compatible with PyQt6/PySide6

jonoomph committed:
Fixing regression in PyQt6 tutorial

jonoomph committed:
Removing selection rectangles from Transitions, Effects, and Emojis (Thumbnail view) - since they are single selection only lists

jonoomph committed:
Protect against setting audio sample rate to None in CheckAudioDevice method

jonoomph committed:
Adding Android-only openshot_shiboken_ext handling for wrapInstance, due to large Arm64 memory addresses (PySide6 forces int() which breaks the pointers in an overflow error)

jonoomph committed:
Protecting setTabOrder calls from potentially using widgets from different windows (quite noisy logs)

jonoomph committed:
Moving to exec() method instead of exec_() when launching OpenShot (if available)

jonoomph committed:
Quite theme miss logs (too noisy at the moment)

jonoomph committed:
Reduce grid size of Emoji list view (they have too much horizontal padding/margin)

jonoomph committed:
Silence more theme misses (when applying theme colors, icons, and styles)

jonoomph committed:
Merge branch 'develop' into qt6-support

# Conflicts:
#	src/windows/export_clips.py
#	src/windows/main_window.py
#	src/windows/preview_thread.py
#	src/windows/title_editor.py
#	src/windows/views/emojis_listview.py
#	src/windows/views/files_listview.py
#	src/windows/views/files_treeview.py
#	src/windows/views/timeline_backend/paint/base.py
#	src/windows/views/timeline_backend/qwidget/base.py
#	src/windows/views/timeline_backend/qwidget/thumbnails.py
#	src/windows/views/timeline_backend/theme.py

jonoomph committed:
Fixing merge regression with QRegEx import for Qt6 path

jonoomph committed:
Fixing more regressions from merging develop into this branch

jonoomph committed:
Another Qt6 regression with emoji dragging

jonoomph committed:
Another Qt6 regression with file adding and dragging

jonoomph committed:
Merge branch 'develop' into qt6-support

jonoomph committed:
Fixing imports on region.py for the Tracker effect, to correctly abstract those imports with qt_api

jonoomph committed:
Fixing a bunch more unwrapped Qt imports and references, so no direct imports are found anywhere else in the code, except for qt_api

jonoomph committed:
Fixing a few Qt6 bugs during testing

jonoomph committed:
Fixing a missing KeyPress, MousebuttonPress qt_api wrapper for QT6

jonoomph committed:
Fixing 2 more Qt6 issues

jonoomph committed:
Scanning code for other similar failures in Qt6

bgermann commented:
You could spare yourself the qt_api.py file by using [QtPy](https://github.com/spyder-ide/qtpy).

jonoomph commented:
> You could spare yourself the qt_api.py file by using [QtPy](https://github.com/spyder-ide/qtpy).

I did not know this exists, lol.  Thanks, I'll take a look.

philocalyst commented:
Does this also migrate to QtWebEngine? @jonoomph 

From my understanding QtWebKit is largely unmaintained.

bgermann commented:
There is already a QtWebEngine codepath additionally to QtWebKit. This PR does not change that.

philocalyst commented:
Is there a need for the dual code paths? Are there users relying on a particular renderer?

▰▰▰▰▰
Miles Wirth 🙃

> From: bgermann ***@***.***>
> Sent: 09 April 2026 21:23
> To: OpenShot/openshot-qt ***@***.***>
> Cc: Miles Wirht ***@***.***>, Comment ***@***.***>
> Subject: Re: [OpenShot/openshot-qt] PyQt5, PyQt6, PySide6 Shim (Switch between different Python bindings for Qt versions) (PR #5885)

>bgermann left a comment (OpenShot/openshot-qt#5885)
>
>There is already a QtWebEngine codepath additionally to QtWebKit. This PR does not change that.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/OpenShot/openshot-qt/pull/5885#issuecomment-4220426933
>You are receiving this because you commented.
>
>Message ID: ***@***.***>

jonoomph committed:
Add new ComfyUI templates for audio denoising/cleanup/repair/voice enhancing (ComfyUI_FL-ClearVoice custom node):
  - Reduce Noise: Reduce hiss, hum, and steady background noise while keeping spoken voice as intact as possible.
  - Enhance Voice: Improve speech clarity and overall voice quality for already usable recordings.
  - Improve Audio Resolution: Improve low-quality or low-resolution audio by using an audio super-resolution model.
  - Repair Audio: Repair degraded, compressed, rough, or damaged audio recordings.

jonoomph committed:
Removing "..." from the end of our new audio denoising context menus, since they do not show a dialog after clicking them.

jonoomph committed:
Merge branch 'develop' into qt6-support

# Conflicts:
#	src/classes/waveform.py
#	src/language/test_translations.py
#	src/tests/test_query.py
#	src/windows/cutting.py
#	src/windows/generate.py
#	src/windows/models/files_model.py
#	src/windows/video_widget.py
#	src/windows/views/find_file.py
#	src/windows/views/timeline_backend/paint/clip.py
#	src/windows/views/timeline_backend/qwidget/base.py
#	src/windows/views/timeline_backend/qwidget/clip.py
#	src/windows/views/timeline_backend/qwidget/thumbnails.py
#	src/windows/views/tutorial.py

jonoomph committed:
Fix timeline property menu QPointF mapping

philocalyst commented:
Or the addition of a compile-time flag? Just makes it impossible to package for Nix as the old engine is marked as vulnerable for 24 CVEs :(

jonoomph committed:
Fixing regression in multi-select files dragging onto timeline (causing duplication of clips)

jonoomph committed:
Fixing a PyQt6 issue on the About/Credits screen, which was erroring due to Regex.

jonoomph committed:
Fixing a tiny issue on a double click handler for PyQt6

jonoomph committed:
Fixing a QFont issue in PyQt6

jonoomph committed:
Fixing a strange crash in pyside6 due to invalid window references on right click context menus

jonoomph committed:
Fixing crashes in pyside6 due to invalid QAction references (or stale references), so attempting a fix.

jonoomph committed:
Fixing crashes in pyside6 testing, and also fixing a regression to filtering on our Credits dialog (developers and supporters)

jonoomph committed:
Fixing font dialog issue in PySide6 with compatibility helper

jonoomph committed:
Integrating new Android override for the preview thread + video widget preset() image flow. This works a bit differently than our desktop versions.

jonoomph committed:
Add new android detect function, some clean up

jonoomph committed:
Fixing file importing for Android platforms (which is quite a bit different than how our desktop works)

jonoomph committed:
Improving "open project" and "save project" to fully support android - required some heavy refactoring in qt_api, to hide some of the ugly android implementation details.

jonoomph committed:
Fixing regression on auto play in preview dialog for this branch - got clobbered along the way

jonoomph committed:
Merge branch 'comfyui-audio-denoising' into qt6-support

jonoomph commented:
We currently support 3 timeline backends: QtWebkit, QtWebEngine, and a custom qwidget built one.  The webview backends will soon be removed - as they are only here for legacy reasons now. I wanted to keep them in the 3.5.x releases - but I plan on removing them from develop branch very soon.

jonoomph committed:
Fixing regression in our translation test script, which imported qt_api and caused a crash.

jonoomph committed:
Fixing a ton of Codacy nitpicks from our Qt6 conversion / refactor

jonoomph committed:
Fixing more Codacy nitpicks - and also a small PySide6 bug on the title editor when text lines change

jonoomph committed:
Fixing small PySide6 bug on the title editor when text lines change (a few more dialogs)

jonoomph committed:
Fixing a few larger Codacy complaints - this is a higher risk set of changes to launch.py

jonoomph committed:
Updating README.md with Qt instructions for switching between PyQt5/PyQt6/PySide6

jonoomph committed:
Adding 2 new audio advanced ai features: Noise -> Reduce/Remove and Clarity -> Speech. Once is a model-based denoiser, and one is a upsampler model - both neural networks, both fast, and both great.

jonoomph committed:
Fixing Windows path regression for assets folder... only content:// paths are treated as Android URIs. Normal local paths, including Windows drive-letter paths, now resolve to the project’s asset folder again. I also added
  focused tests for both Windows local paths and content:// URIs.

jonoomph committed:
Fixing regression when copy/paste was not correctly taking the vertical scroll position of the timeline into account, and pasting clips in the wrong track (not the track selected)

jonoomph committed:
Fixing a bug where copy/pasting a clip at the right edge of the timeline did not extend the project's duration to include the newly pasted clips.

jonoomph committed:
Fix bug where CTRL+V does not paste a clip/transition on the correct track when vertical scrolling is involved on the timeline.

jonoomph committed:
Fixing another CTRL+V paste bug for transitions, which would prevent them from being pasted a second time.
